src/dash-table/components/ControlledTable/index.tsx

```diff
@@ -806,7 +837,67 @@ export default class ControlledTable extends PureComponent<ControlledTableProps>
                     <button className='next-page' onClick={this.loadNext}>Next</button>
                 </div>
             )}
-            <ExportButton {...buttonProps} />
+        </div>);
+    }
+
+    renderMenu() {
+        if (!this.showToggleColumns) {
+            return null;
+        }
+
+        const {
+            activeMenu,
+            columns,
+            hidden_columns,
+            merge_duplicate_headers,
+            setState
+        } = this.props;
+
+        const labelsAndIndices = this.labelsAndIndices(columns, columns, merge_duplicate_headers);
+        const lastRow = labelsAndIndices.length - 1;
+
+        return (<div
+            className='dash-spreadsheet-menu-item'
+            ref={this.menuRef}
+        >
+            <button className='show-hide'
+                onClick={() => setState({
+                    activeMenu: activeMenu === 'show/hide' ? undefined : 'show/hide'
+                })}
+            >Toggle Columns</button>
+            {activeMenu !== 'show/hide' ?
+                null :
+                <div className='show-hide-menu'>
+                    {R.unnest(labelsAndIndices.map(([, indices], i) => indices.map((index, j) => {
+                        const spansAllColumns = indices.length === 1;
+                        const column = columns[index];
+
+                        const checked = !hidden_columns || hidden_columns.indexOf(column.id) < 0;
+                        const hideable = getColumnFlag(i, lastRow, column.hideable);
+
+                        const disabled = (spansAllColumns && checked) || (!hideable && checked);
+
+                        return {
+                            i: index,
+                            j,
+                            component: !hideable ? null : (<div className='show-hide-menu-item'>
+                                <input
+                                    type='checkbox'
+                                    checked={checked}
+                                    disabled={disabled}
+                                    onClick={this.toggleColumn.bind(this, column, i, merge_duplicate_headers)}
+                                />
+                                <label>{!column.name ?
+                                    column.id :
+                                    typeof column.name === 'string' ?
+                                        column.name :
+                                        column.name.slice(0, i + 1).filter(name => name.length !== 0).join(' | ')
+                                }</label>
+                            </div>)
+                        };
+                    }))).filter(i => !R.isNil(i)).sort((a, b) => a.j - b.j).sort((a, b) => a.i - b.i).map(a => a.component)}
```

Comment:
This is a bit messy -- can rework if seen as problematic.

Generates a sparse component[][] containing the list of inputs but keeping their associated index, remove nil items and sort in reverse order of condition/priority (columns are more important than rows -- ordering is like a radix sort), retrieve the component from the wrapper.

Reply:
`.sort((a, b) => (a.i - b.i) || (a.j - b.j))` ?

Reply:
Right. That should work too. Checking / updating.